server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateService.java

```diff
@@ -991,12 +992,13 @@ private static void validateDataStreamsStillReferenced(
         };
 
         // Find data streams that are currently unreferenced
-        final Set<String> currentlyUnreferenced = findUnreferencedDataStreams.apply(project);
+        final Set<String> currentlyUnreferenced = findUnreferencedDataStreams.apply(project.templatesV2());
 
-        // Generate a metadata as if the new template were actually in the cluster state
-        final ProjectMetadata updatedMetadata = ProjectMetadata.builder(project).put(templateName, newTemplate).build();
+        // Generate a map as if the new template were actually in the cluster state
+        final var updatedTemplatesMap = new HashMap<>(project.templatesV2());
+        updatedTemplatesMap.put(templateName, newTemplate);
```

Comment:
Copying this map is still not super cheap if there are lots and lots of composable index templates, but it's definitely a lot cheaper than building a whole `ProjectMetadata`. I don't think this is really worth worying about.

Reply:
Also, doing this check (`validateDataStreamsStillReferenced`) is redundant in some cases, like when we only update mappings of a component or index template. I think we could only do this check if either the hidden setting or the index patterns changed, but I'm not sure if the added complexity is worth it.